File: src/leetcode_javascript/Combination_Sum_39.js
/**
 *
 * 39 Combination Sum
 *
 * Given a set of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.

  The same repeated number may be chosen from C unlimited number of times.

  Note:
  All numbers (including target) will be positive integers.
  The solution set must not contain duplicate combinations.
  For example, given candidate set [2, 3, 6, 7] and target 7,
  A solution set is:
  [
    [7],
    [2, 2, 3]
  ]
 *
 *
 * https://leetcode.com/problems/combination-sum/
 */

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    if(candidates.length < 1 || candidates === null) return [];
    var result = [];
    var temp = [];
    candidates.sort((a, b) => a - b);
    dfs(0, target);
    return result;

    function dfs(startIdx, target) {
       if(target === 0) result.push(temp.slice());
       if(target < 0) return;
       if(startIdx === candidates.length) return;
       for(var i = startIdx; i < candidates.length; i++) {
           temp.push(candidates[i]);
           dfs(i, target - candidates[i]);
           temp.pop();
       }
    }
};

File: src/leetcode/dfs/Combination_Sum_39.java
package leetcode.dfs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by sunbo_000 on 10/17/2016.
 */
public class Combination_Sum_39 {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        if (candidates == null || candidates.length == 0) return result;
        Arrays.sort(candidates);
        dfs(candidates, 0, target, new ArrayList<>(), result);
        return result;
    }

    private void dfs(int[] candidates, int pointer, int target, List<Integer> temp, List<List<Integer>> result) {


        if (target == 0) {
            result.add(new ArrayList<>(temp));
            return;
        }
        for (int i = pointer; i < candidates.length; i++) {
            if (target < candidates[i]) return;
            temp.add(candidates[i]);
            dfs(candidates, i, target - candidates[i], temp, result);
            temp.remove(temp.size() - 1);
        }
    }

    public static void main(String[] args) {
        Combination_Sum_39 solution = new Combination_Sum_39();
        int[] candidates = {1, 2};
        int target = 4;
        List<List<Integer>> result = solution.combinationSum(candidates, target);
    }
}
